File: 算法/代码随想录算法PDF/00调试源代码/51.n-皇后.py
#
# @lc app=leetcode.cn id=51 lang=python3
#
# [51] N 皇后
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        path = [['.'] * n for _ in range(n)]
        def backtracking(row):
            nonlocal n
            if row == n:
                line = []
                for tmp in path:
                    line.append(''.join(tmp))
                res.append(line)
                return
            for col in range(n):
                if isValid(row, col):
                    path[row][col] = 'Q'
                    backtracking(row + 1)
                    path[row][col] = '.'
        def isValid(row, col):
            nonlocal n
            for i in range(row):
                if path[i][col] == 'Q': return False
            i = row - 1
            j = col - 1
            while i >= 0 and j >= 0:
                if path[i][j] == 'Q': return False
                i -= 1
                j -= 1
            i = row - 1
            j = col + 1
            while i >= 0 and j < n:
                if path[i][j] == 'Q': return False
                i -= 1
                j += 1
            return True
        
        backtracking(0)
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/51.n-皇后.cpp
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N 皇后
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> res;  // 结果
    vector<string> path;  // 盘面排布
    int num;  // 盘面大小
    bool isValid(int row, int col) {
        for (int i = row - 1; i >= 0; i--) {
            if (path[i][col] == 'Q') return false;
        }
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--){
            if (path[i][j] == 'Q') return false;
        }
        for (int i = row - 1, j = col + 1; i >= 0 && j < num; i--, j++){
            if (path[i][j] == 'Q') return false;
        }
        return true;
    }
    void backtracking(int row) {
        if (row == num) {
            res.push_back(path);
            return ;
        }
        for (int i = 0; i < num; i++) {
            if (isValid(row, i)) {  // 合法可以放
                path[row][i] = 'Q';
                backtracking(row + 1);  // 递归
                path[row][i] = '.';  // 回溯
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
        res.clear();
        num = n;
        path.assign(n, string(n, '.'));
        backtracking(0);
        return res;
    }
};
// @lc code=end

